### ISSUE TITLE
Postprocess Analysis is double counting top symbol from resultsIterator because of duplication from ChoiceIterator

### ISSUE THREAD

**Issue by alandtse** [openalpr/openalpr#261]
This may be as designed, but Postprocess Analysis (postprocess.cpp) appears to give significant preference for the first result returned by Tesseract's resultsIterator (used in ocr.cpp) because the same symbol will be returned by choiceIterator with the same score and added twice.

For example, in the below the correct value is "O", but the result of the three thresholds is weighing "0" with twice as much weight because it'll appear again in choiceIterator results.

>charpos2 line0: threshold 0:  symbol 0, conf: 93.227600 font: washington (index 5) size 11px	- O conf: 93.493355
>		 	- 0 conf: 93.227600
>		 	- Q conf: 91.606186
>		 	- D conf: 87.966324
>		 	- G conf: 79.337433
>		 	- B conf: 78.775238

>---------------------------------------------
>charpos2 line0: threshold 1:  symbol Q, conf: 91.510323 font: fl (index 6) size 11px	- Q conf: 91.510323
>		 	- O conf: 90.865555
>		 	- 0 conf: 88.982857
>		 	- D conf: 86.211647
>		 	- U conf: 79.387970
>		 	- G conf: 77.661942

>---------------------------------------------
>charpos2 line0: threshold 2:  symbol 0, conf: 88.421432 font: california (index 4) size 11px	- O conf: 90.863953
>		 	- Q conf: 89.370117
>		 	- 0 conf: 88.421432
>		 	- D conf: 86.675827

>---------------------------------------------
>PostProcess Line 0 Letter: 2 0 -- score: 127.281 -- occurrences: 5
>PostProcess Line 0 Letter: 2 Q -- score: 103.997 -- occurrences: 4
>PostProcess Line 0 Letter: 2 O -- score: 80.2229 -- occurrences: 3
>PostProcess Line 0 Letter: 2 D -- score: 65.8538 -- occurrences: 3
>PostProcess Line 0 Letter: 2 G -- score: 26.9994 -- occurrences: 2
>PostProcess Line 0 Letter: 2 U -- score: 14.388 -- occurrences: 1
>PostProcess Line 0 Letter: 2 B -- score: 13.7752 -- occurrences: 1
>PostProcess Line 0 Letter: 2 ~ -- score: 4.83742 -- occurrences: 4

One reason this may not be a desirable result is that Tesseract (at least in 2012) did not guarantee the resultsIterator returns the symbol with the highest score of all the results from choiceIterator because it may rely on built-in rules based on neighboring characters.  Tesseract acknowledged that in a bug report by someone doing license plate recognition:
>There are built-in rules about sensible combinations of characters that make up a word that are helping out here, even though it is not a dictionary word.
https://code.google.com/p/tesseract-ocr/issues/detail?id=775 

I have implemented a fix where a duplicate entry is not added and the correct scoring and selection is made with this specific example.  Here's the new log:
>PostProcess Line 0 Letter: 2 O -- score: 80.2229 -- occurrences: 3
>PostProcess Line 0 Letter: 2 Q -- score: 77.4866 -- occurrences: 3
>PostProcess Line 0 Letter: 2 0 -- score: 75.6319 -- occurrences: 3
>PostProcess Line 0 Letter: 2 D -- score: 65.8538 -- occurrences: 3
>PostProcess Line 0 Letter: 2 G -- score: 26.9994 -- occurrences: 2
>PostProcess Line 0 Letter: 2 U -- score: 14.388 -- occurrences: 1
>PostProcess Line 0 Letter: 2 B -- score: 13.7752 -- occurrences: 1
>PostProcess Line 0 Letter: 2 ~ -- score: 4.83742 -- occurrences: 4

**Comment by matthill** [comment]
Awesome catch!  That's been in the code since the beginning, and has gone unnoticed.

If you want to post a patch, I'll run it through a benchmark and merge it.

**Comment by alandtse** [comment]
Hi Matt,
Before I provide a patch, can you please explain your plans for the commercial vs open source version?  I saw that contributions are under CLA and I get that it lets you offer a commercial version without AGPL obligations and also makes it easier to get acquired if you do real well (and all the power to you if that happens).*  

I was wondering if you were planning to set a policy where you will back-port some of the commercial features to the open source version after a certain amount of time or version. I love the idea that openalpr is open source but can already see some features in the open source version falling behind the commercial version (e.g., state detection, video processing, etc.). It'd be cool if they were in sync, but I get that the Red Hat commercial support model is not for everyone.

*Full disclosure, I'm a lawyer who does open source licensing for my day job, including advising on open source acquisitions so I have a habit of reading CLAs fully. ;)

**Comment by matthill** [comment]
Hey Alan,

It's great to be speaking to an expert on the topic.  Please let me know if I'm off-base on any of my laymen legal assumption.

My opinion is that CLAs should be attached to every open source project unless it's under a permissive (e.g., MIT, BSD, etc.) license.  I think it's important for some entity to have ownership so that the IP could be relicensed.

For example, if I no longer had any commercial or personal interest in the project, without a CLA, the code would be locked into AGPL forever.  With the CLA, I could decide to relicense it under MIT.  On the flip side, I could cease all development on the open source project and make my further enhancements proprietary.  But, the AGPL code is still out there (I can't legally take it back) and it could continue to grow via community development.

Generally, my plan is to keep all the core license plate recognition algorithms exactly the same between commercial and open source.  That is how it is now.  Anything that is only in the commercial code (e.g., state detection, video processing, real-time stream processing) is modularized to allow open source contributions to add the functionality.  I don't reject any good contributions, regardless of whether I have a commercial implementation.

**Comment by alandtse** [comment]
Yes your general understanding is correct, CLAs are primarily to allow a project to change its licensing in the future and realistically is a best practice.  It's generally best practice even if you're using permissive licenses because it gives the project control moving forward (and people who buy out start-ups want this control).  Any older version under AGPL will forever be out there but you could move forward under the new license.

Regarding your current practice, the modularization would likely only help as long as you have been diligent about only taking code under CLA or permissive licenses. If you started taking in any GPL or other copyleft licenses (including AGPL from forked copies of openalpr), you likely would be under copyleft provisions for your commercial code because of that third-party copyleft code.  So if you'd like to maintain the modularization, you should be careful there.

Coming back to the current issue, I asked primarily because I wasn't sure how I felt about openalpr's use of CLAs for an AGPL project. Generally for copyleft projects, CLAs just give the project owner the right to take contributions and put them in the commercial version without ever providing commercial features back; the project owner isn't subject to the copyleft everyone else is under. So it gave me pause to see that some stuff I particularly wanted to work on improving (state detection, video processing) was effectively being kept worse in the open source project. In sum, it felt a bit unfair.

But to get off my soap box, I'll likely put in a pull request just for this bug fix and work on my separate fork for other improvements (subject of course to AGPL). I'll plan to submit bug reports if I see them but I think patches will be on a case-by-case basis. Again, I totally get why you use a CLA, but I wanted to at least let you know my thoughts since you may not have been exposed to this type of viewpoint before. If you ever decide in the future to adopt some policy or pledge that the commercial version improvements eventually make it back, I'd be more than happy to help you with that.

**Comment by matthill** [comment]
Thanks, Alan.  I appreciate the candid advice, and the patch.  Yes, I do only use code under CLA or permissive license.  I release the majority of my work under the core open source project, and will continue to do so.   It's a bit of a compromise between OS and commercial.  If there weren't a commercial component to this project, development would stall.  

I ran it through my single-image benchmarks (https://github.com/openalpr/benchmarks/tree/master/endtoend).  I see the following results.  I'm going to hold onto this PR until I can get the accuracy to be the same.  I'd like to accept this patch, and then explicitly add the top result twice.  At least we need to be explicit about it rather than letting it be a happy accident.  It seems to make a significant difference wrt to accuracy.


US Before:

Accuracy scores (higher is better)
Percent of plates DETECTED: 97.2973
Percent of correct TOP10:   69.8198
Percent of correct MATCHES: 50.9009

False Positives Score (lower is better)
False DETECTIONS per image: 0.211712
False RESULTS per image:    0.256757

US After patch:

Accuracy scores (higher is better)
Percent of plates DETECTED: 97.2973
Percent of correct TOP10:   66.2162
Percent of correct MATCHES: 46.3964

False Positives Score (lower is better)
False DETECTIONS per image: 0.211712
False RESULTS per image:    0.288288

EU Before:

Accuracy scores (higher is better)
Percent of plates DETECTED: 99.0741
Percent of correct TOP10:   88.8889
Percent of correct MATCHES: 67.5926

False Positives Score (lower is better)
False DETECTIONS per image: 0.101852
False RESULTS per image:    0.111111


EU After:
Accuracy scores (higher is better)
Percent of plates DETECTED: 99.0741
Percent of correct TOP10:   89.8148
Percent of correct MATCHES: 53.7037

False Positives Score (lower is better)
False DETECTIONS per image: 0.101852

**Comment by alandtse** [comment]
You've probably been living with the double counting so long that your test cases benefit from having the double count.  I think the edge case will be probably between O/0 and maybe Q.

When you do add the top score again, I think doubling the top ChoiceIterator option may be better since the top ResultsIterator is dependent on some undefined Tesseract rules. You may be able to optimize by not doing ResultsIterator and going straight to ChoiceIterator, but I'm not sure if ChoiceIterator is dependent on ResultsIterator being run.

**Comment by matthill** [comment]
Yes, you may be right.  Although I did test it on a throwaway database, and I saw similar accuracy loss.  It could also be that other parts of the software are tuned towards expecting that double counted data as well.
